stlab/concurrency/router.hpp

```diff
@@ -0,0 +1,137 @@
+/*
+    Copyright 2015 Adobe
+    Distributed under the Boost Software License, Version 1.0.
+    (See accompanying file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+*/
+
+/**************************************************************************************************/
+
+#ifndef STLAB_CONCURRENCY_ROUTER_HPP
+#define STLAB_CONCURRENCY_ROUTER_HPP
+
+#include <algorithm>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include <stlab/concurrency/channel.hpp>
+
+/**************************************************************************************************/
+
+namespace stlab {
+
+/**************************************************************************************************/
+
+inline namespace v1 {
+
+/**************************************************************************************************/
+
+template <class T, class K>
+class router {
+    using channel_pair = std::pair<stlab::sender<T>, stlab::receiver<T>>;
+    using route_pair = std::pair<K, channel_pair>;
```

Comment:
The type K should be const as the Key is const in std::map, etc.

Reply:
If we do that we can't sort the `std::vector<route_pair>`. If the user provides a list of all possible routes on construction then we don't need to sort and we can accomplish all your other asks (except the `std::unique_ptr<>`). Thoughts?

Reply:
I went ahead and made a version that provides the routes on construction. I added the test program to the PR comment.